### ISSUE TITLE
getData doesn't send current data / state as a second argument

### ISSUE THREAD

**Issue by burgalon** [klis87/redux-requests#332]
While migrating from 0.26 => 0.27 and using the  actions meta getData, I would like to use the current state.data to merge some attributes together with the current response. For example, pagination (infinite scroll style) which adds more to the bottom of the list.
In the past, `requestReducer` `getData: (state, action)` signature. With 0.27 `getData(data)` does not allow it....
  
I'm trying to use `mutations` instead to get the former `data` but for some reason it's not being called?!

**Comment by klis87** [comment]
@burgalon you want to merge local state with remote state?

**Comment by klis87** [comment]
anyway, getData works, see basic example https://github.com/klis87/redux-saga-requests/blob/master/examples/basic/src/store/actions.js#L11

**Comment by burgalon** [comment]
the use case is somewhat similar to the example with cache
```
const fetchBook = id => ({
  type: FETCH_BOOK,
  request: { url: `/books/${id}`},
  meta: {
    cache: true,
    requestKey: id,
    requestsCapacity: 2 // optional, to clear old cache exceeding this number
  },
});```

but in my case i need a more elaborate merging of the data and former data

**Comment by burgalon** [comment]
also for now I wouldn't like to use the redux-saga-requests abstraction for cache as I have some current logic in place of when to fetch and when to use current redux state... for example a "Force refresh" button

**Comment by burgalon** [comment]
What do you mean by `getData works` ?
I don't sure I understand the example:
```
export const fetchPost = id => ({
  type: FETCH_POST,
  request: [{ url: `/posts/${id}` }, { url: `/posts/${id}/comments` }],
  meta: {
    getData: data => ({
      ...data[0],
      comments: data[1],
    }),
  },
});
```

What is data[0] and data[1]?  this still doesn't seem to answer the need to merge (cache) the old posts...

**Comment by burgalon** [comment]
nm ... i get it now.. it's an array of request... :)
However still need previous data / state

**Comment by klis87** [comment]
@burgalon but is there any reason getData doesnt work? pls give example which could make me understand

also you can use so called local mutations to update data in any way you like

you can also use reselect to map data to sth else, like in https://github.com/klis87/redux-saga-requests#reducers-arrow_up

**Comment by klis87** [comment]
ah, I get you know, probably you want to merge previous data with new one, you could also have things like:

const fetchBooks = () => ({
  type: FETCH_BOOKS,
  request: { url: '/books' },
});

const mergeBooks = () => ({
  type: MERGE_BOOKS,
  request: { url: '/books' },
  meta: {
    asMutation: true,
    mutations: {
      [FETCH_BOOKS]: //.....
    },
  }
});

**Comment by klis87** [comment]
above should solve things for you, but I can understand that maybe getData should have (newData, currentData) => newData signature

**Comment by burgalon** [comment]
yap!

**Comment by burgalon** [comment]
but for some reason the mutations callback is not called :(

**Comment by burgalon** [comment]
in debugger I think it's not called because the `queries` key is never initialized once `asMutation` is used. So `.filter(actionType => !!state.queries[actionType])` is filtering it

**Comment by burgalon** [comment]
just to be clear... in my case I have only `mergeBooks` and not `fetchBooks`

**Comment by burgalon** [comment]
example use case:
```
export const fetchStructuredSettings = (key: string) => ({
  type: FETCH_STRUCTURED_SETTINGS,
  request: {
    method: 'get',
    service: 'config/v1',
    url: `settings/${key}/structured`,
  },
  meta: {
    takeLatest: false,
    asMutation: true,
      mutations: {
      [FETCH_STRUCTURED_SETTINGS]: (data: any, mutationData: any) => {
        debugger;
        return {
          ...data,
          [mutationData.data.key]: mutationData.data.value || SETTINGS_DEFAULTS[mutationData.data.key],
        }
      },
    },
  },
});
```

**Comment by burgalon** [comment]
doesn't it make sense for example to have `createPost` action which mutations the `FETCH_POSTS` and adds the new post to the data ? In this case `fetchPosts()` might never have been called, and then the mutation would not happen

**Comment by klis87** [comment]
@burgalon In order for this to work, you need to have query, only queries can hold data, mutations doesnt have data stored.

What you could do is this:
```js
export const fetchStructuredSettings = (key: string, initial: boolean) => ({
  type: FETCH_STRUCTURED_SETTINGS,
  request: {
    method: 'get',
    url: `settings/${key}/structured`,
  },
  meta: {
    takeLatest: false,
    asMutation: !initial,
      mutations: {
      [FETCH_STRUCTURED_SETTINGS]: (data: any, mutationData: any) => {
        return {
          ...data,
          [mutationData.data.key]: mutationData.data.value || SETTINGS_DEFAULTS[mutationData.data.key],
        }
      },
    },
  },
});
```
The point is this, you can have a mutation and query with the same type, but a mutation can mutate data of a query which already exists, hence you need asMutation: !initial, for the first time you will run it as query, next dispatches can be run as mutations so you will be able to merge data

**Comment by burgalon** [comment]
that's understood, it's a sort of a hack for `getData` not getting second argument of existing `data`. Can I create a PR to add that?

**Comment by klis87** [comment]
@burgalon yes, thx! I think this is a good use case for infinite scroll etc, when you want to store everything inside 1 query. It wont hurt for sure :)

**Comment by burgalon** [comment]
from which branch should I fork?

**Comment by klis87** [comment]
`removing-sagas`, provided you will be patient for me to merge as this branch is huge, could take me 1-2 weeks more. You could use hack we described for now.

**Comment by burgalon** [comment]
which hack? the hack of `mutations` does not work because of `.filter(actionType => !!state.queries[actionType])` filtering in queriesReducer

**Comment by burgalon** [comment]
is the current 0.27 released from master? Can I for the PR from that?

**Comment by klis87** [comment]
it will work with  `asMutation: !initial`, if you call if once as query, then query will exist and .filter(actionType => !!state.queries[actionType]) will spot this query.

if this pr will be very small which I believe it will be, lets do it from master, it wont cause me any headaches probably :)

**Comment by burgalon** [comment]
for the `initial` hack to work, I also need to alternate the reducer part between the `mutations` and the `getData`... this is pretty nasty

**Comment by burgalon** [comment]
argg.... the PR doesn't look easy. Didn't realize that `getData` is now called from `sendRequest` and not from a reducer and doesn't have easy access to the previous state

**Comment by klis87** [comment]
if it is not easy, despite the fact it is nasty, maybe lets stick to hack for a while, I will do it myself fast before next release, in next release sendRequest wont even exists as sagas will be removed, so it would be a pity to waste time on this.

**Comment by klis87** [comment]
@burgalon closed in `0.28.0`, now you have currentData as 2nd arg passed to `getData`, you can also use meta.onSuccess and do whatever you want, as you have store access there and can return whatever you want
